File: Level(2)target/Level(2)target.js
function solution(numbers, target) {
    var answer = 0;
    var answer = 0;

    let root = new BinarySearchTree();
    root.insert(0);
    numbers.forEach(function(val) {
       root.insert(val); 
    });

    answer = root.DFSPreOrder(target);
    return answer;
}

class Node {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }   
}
class BinarySearchTree {
    constructor() {
        this.root = null;
    }
insert(value) {
        let newNode = new Node(value);
        if(this.root === null) {
            this.root = newNode;
            return this;
        } else {
            let current = this.root;
            function traverse(node) {
                if(node.left) traverse(node.left);
                if(node.right) traverse(node.right);
                if(node.left === null) {
                    let leftNode = new Node(-value);
                    let rightNode = new Node(value);
                    node.left = leftNode;
                    node.right = rightNode;
                }
            }
            traverse(current);
            return this;
        }
    }

File: Level(2)target/Level(2)target.cpp
#include <string>
#include <vector>

using namespace std;

bool promising(int depth, const vector<int> &numbers, int search, int target) {
    int lastSum = 0;
    for (int i = depth; i < numbers.size(); i++) {
        lastSum += numbers[i];
    }

    if (search + lastSum < target || search - lastSum > target) {
        return false;
    }
    else return true;
}

// void searchNum(int &answer, int depth, const vector<int> &numbers, int search, int target) {
//     if (promising(depth, numbers, search, target)) {
//         if (depth == numbers.size()) {
//             answer++;
//             return;
//         }

//         search += numbers[depth];
//         searchNum(answer, depth + 1, numbers, search, target);

//         search -= (2 * numbers[depth]);
//         searchNum(answer, depth + 1, numbers, search, target);
//     }
// }

// int solution(vector<int> numbers, int target) {

//     int answer = 0;

//     searchNum(answer, 0, numbers, 0, target);

//     return answer;
// }
